cmd/maya-apiserver/spc-watcher/select_disk.go

```diff
@@ -0,0 +1,270 @@
+// to efficiently select the nodes and disk for dynamic pool provisioning.
+type nodeDisk struct {
+	// diskCount is the count of usable disks that can be used in storagepool provisioning.
+	diskCount int
+	//diskList is the list of usable disks that can be used in storagepool provisioning.
+	diskList []string
+}
+
+// getDiskList is a wrapper function which will receive list of disks
+// that can be used for dynamic pool provisioning at runtime.
+// The function finally returns the disk list to the caller (i.e. getCasPoolDisk function).
+func getDiskList(cp *v1alpha1.CasPool) ([]string, error) {
+
+	// Get kubernetes clientset
+	// namespaces is not required, hence passed empty.
+	newK8sClient, err := k8s.NewK8sClient("")
+
+	if err != nil {
+		return nil, err
+	}
+	// Get openebs clientset using a getter method (i.e. GetOECS() ) as
+	// the openebs clientset is not exported.
+	newOecsClient := newK8sClient.GetOECS()
+
+	// Create instance of clientset struct defined above which binds
+	// ListDisk method and fill it with openebs clienset (i.e.newOecsClient ).
+	newClientSet := clientSet{
+		oecs: newOecsClient,
+	}
+	// nodeDiskAlloter will try to return a list of disks so that maxpool number of storagepool
+	// is provisioned.
+	diskList, err := newClientSet.nodeDiskAlloter(cp)
+	if err != nil {
+		return nil, err
+	}
+
+	return diskList, nil
+}
+
+// nodeDiskAlloter will try to allot nodes for pool creation as specified in
+// maxPool field of the storagepoolclaim.
+
+// For exapmle, if maxPool=5 and minPool=3, it will try to search for 5 nodes that will qualify for
+// pool provisioning. At least 3 node should qualify else pool will not be provisioned and pool creation
+// will be aborted gracefully with proper log messages.
+
+// If no minPool field is present,at least one node must qualify for pool provisioning.
+
+// nodeDiskAlloter can be made more intelligent as per the required pool constraints for alloting nodes.
+func (k *clientSet) nodeDiskAlloter(cp *v1alpha1.CasPool) ([]string, error) {
+
+	// pendingAllotment holds the value for the remaining node allotments
+	// for pool provisioning.
+	var pendingAllotment int
+
+	// assign maxPools to pendingAllotment as right now maxPool is the number of allotments
+	// that needs to be done.
+	pendingAllotment = cp.MaxPools
+	// get the labels on the basis of which disk list will be filtered
+	diskFilterLabel := diskFilterConstraint(cp.Type)
+
+	// Request kube-apiserver for the list of disk (powered by NDM)
+	// Currently, all the disks are returned,but the disk that is already a part of pool
+	// should not be returned.
+	listDisk, err := k.oecs.OpenebsV1alpha1().Disks().List(mach_apis_meta_v1.ListOptions{LabelSelector: diskFilterLabel})
+	if err != nil {
+		return nil, fmt.Errorf("error in getting the disk list:%v", err)
+	}
+	if listDisk == nil {
+		return nil, errors.New("no disk object found")
+	}
+
```

Comment:
Should we also check for len(listDisk) > 0 here?

Reply:
Following is better,if  we want to check len: 
```
if len(listDisk.Items) == 0 {
		return nil, errors.New("no disk object found")
	}
```
Will update it.

Reply:
fixed